template/src/components/Toast.tsx

```diff
@@ -1,202 +1,175 @@
-  messageContainer: () => ViewProps
-  textStyle: TextProps
-}
+const messageContent = {
+  [TOAST_TYPE.SUCCESS]: {
+    color: colors.success,
+  },
+  [TOAST_TYPE.ERROR]: {
+    color: colors.error,
+  },
+  [TOAST_TYPE.INFO]: {
+    color: colors.info,
+  },
+  [TOAST_TYPE.WARNING]: {
+    color: colors.warning,
+  },
+} as const
 
-enum EToastType {
-  SUCCESS = 'success',
-  ERROR = 'error',
-  INFO = 'info',
+const initState = {
+  message: '',
+  type: TOAST_TYPE.INFO,
+  subMessage: '',
 }
 
-export class Toast extends React.PureComponent<IToastProps, IToastState> {
-  frameID(frameID: any) {
-    throw new Error('Method not implemented.')
-  }
-  offset: Animated.Value
-  opacity: Animated.Value
-  animated: Animated.AnimatedProps<any> | null
-  // Static methods
-  static success(text: string) {
-    Emitter.emit('SHOW_TOAST_MESSAGE', {message: text, type: EToastType.SUCCESS})
-  }
-
-  static error(text: string) {
-    Emitter.emit('SHOW_TOAST_ERROR', {message: text, type: EToastType.ERROR})
-  }
-
-  static info(text: string) {
-    Emitter.emit('SHOW_TOAST_INFO', {message: text, type: EToastType.INFO})
-  }
-
-  constructor(props: IToastProps, IState: IToastState, offset: any) {
-    super(props)
-    this.state = {
-      message: '',
-      type: EToastType.SUCCESS,
-    }
-    this.offset = new Animated.Value(-HEIGHT)
-    this.opacity = new Animated.Value(0)
-    this.animated = null
-  }
-
-  componentDidMount() {
-    Emitter.on('SHOW_TOAST_MESSAGE', this.displayMessage)
-    Emitter.on('SHOW_TOAST_ERROR', this.displayMessage)
-    Emitter.on('SHOW_TOAST_INFO', this.displayMessage)
-  }
-
-  componentWillUnmount() {
-    Emitter.rm('SHOW_TOAST_MESSAGE')
-    Emitter.rm('SHOW_TOAST_ERROR')
-    Emitter.rm('SHOW_TOAST_INFO')
-  }
-
-  displayMessage = ({message, type}: IToastState): void => {
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    // @ts-ignore
-    window.cancelAnimationFrame(this.frameID)
-
-    this.offset.setValue(HEIGHT * -1)
-    this.setState({message, type})
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    // @ts-ignore
-    this.frameID = window.requestAnimationFrame(() => {
-      this.animated = Animated.sequence([
-        Animated.delay(100),
+export const Toast: React.FC = () => {
+  const insets = useSafeAreaInsets()
+  const [state, setState] = useState<IToastState>(initState)
+  const HEIGHT = insets.bottom + metrics.toast
+  const animationRef = useRef<Animated.CompositeAnimation>()
+  const animation = useRef(new Animated.Value(0)).current
+  const offset = animation.interpolate({
+    inputRange: [0, 1],
+    outputRange: [HEIGHT, 0 - metrics.xxs], // padding bottom metrics.xxs
+  })
+  const opacity = animation.interpolate({
+    inputRange: [0, 1],
+    outputRange: [0, 1],
+  })
+
+  const displayMessage = (args: IToastState): void => {
+    animation.setValue(0)
+    setState(args)
+    animationRef.current?.stop()
+    setTimeout(() => {
```

Comment:
Do we need to use setTimeout?

Reply:
Yes, we need the animation to completely stop after calling `animationRef.current?.stop()` to execute again

Reply:
Then why didn't you use it here?
https://github.com/saigontechnology/rn-base-project-typescript/pull/217/files#diff-3811f7c80477ce2277c50dccb8088d5a7a44148774970c2cdfdb3a81e56cae63R92

Reply:
This place doesn't need it, to make things clear, there are 2 cases:
- Simply dismiss the toast => the animation will stop and revert the value to default
- Call toast when another toast is already in display => this case we will have to wait for animation to completely stop, then call the next toast animation (or else the animation will not work properly) => So we need to setTimeout here and not on `dismiss` action